File: phases/15-systems-performance/08-vectorization-in-practice-auto-and-intrinsics/code/main.cpp
#include <immintrin.h>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <numeric>
#include <vector>

#ifdef _MSC_VER
#define ALIGNED_ALLOC(align, size) _aligned_malloc(size, align)
#define ALIGNED_FREE(ptr) _aligned_free(ptr)
#else
#define ALIGNED_ALLOC(align, size) std::aligned_alloc(align, size)
#define ALIGNED_FREE(ptr) free(ptr)
#endif

static constexpr int N = 1024 * 1024;
static constexpr int ITERS = 200;

static float* alloc_aligned(int count, int alignment = 64) {
    size_t bytes = static_cast<size_t>(count) * sizeof(float);
    float* ptr = static_cast<float*>(ALIGNED_ALLOC(alignment, bytes));
    for (int i = 0; i < count; i++) ptr[i] = 0.0f;
    return ptr;
}

static void fill_random(float* data, int n) {
    for (int i = 0; i < n; i++) {
        data[i] = static_cast<float>(rand()) / static_cast<float>(RAND_MAX) - 0.5f;
    }
}

// ─── Scalar implementations ───

float dot_scalar(const float* a, const float* b, int n) {
    float sum = 0.0f;
    for (int i = 0; i < n; i++) {
        sum += a[i] * b[i];
    }
    return sum;
}

float sum_scalar(const float* data, int n) {
    float s = 0.0f;
    for (int i = 0; i < n; i++) {
        s += data[i];
    }
    return s;
}

int filter_scalar(const float* src, float* dst, int n, float threshold) {
    int count = 0;
    for (int i = 0; i < n; i++) {
        if (src[i] > threshold) {
            dst[count++] = src[i];
        }
    }
    return count;
}

// ─── Auto-vectorizable implementations ───
// __restrict__ tells the compiler a and b don't alias, enabling vectorization.

float dot_auto(const float* __restrict__ a,
               const float* __restrict__ b, int n) {
    float sum = 0.0f;
    for (int i = 0; i < n; i++) {
        sum += a[i] * b[i];
    }
    return sum;
}

float sum_auto(const float* __restrict__ data, int n) {
    float s = 0.0f;
    for (int i = 0; i < n; i++) {
        s += data[i];
    }
    return s;
}

// ─── SSE intrinsics (128-bit, 4 × float32) ───

float dot_sse(const float* a, const float* b, int n) {
    __m128 sum_vec = _mm_setzero_ps();
    int i = 0;
    for (; i + 3 < n; i += 4) {
        __m128 va = _mm_loadu_ps(a + i);
        __m128 vb = _mm_loadu_ps(b + i);
        sum_vec = _mm_add_ps(sum_vec, _mm_mul_ps(va, vb));
    }
    float tmp[4];
    _mm_storeu_ps(tmp, sum_vec);
    float sum = tmp[0] + tmp[1] + tmp[2] + tmp[3];
    for (; i < n; i++) sum += a[i] * b[i];
    return sum;
}

float sum_sse(const float* data, int n) {
    __m128 sum_vec = _mm_setzero_ps();
    int i = 0;
    for (; i + 3 < n; i += 4) {
        __m128 v = _mm_loadu_ps(data + i);
        sum_vec = _mm_add_ps(sum_vec, v);
    }
    float tmp[4];
    _mm_storeu_ps(tmp, sum_vec);
    float sum = tmp[0] + tmp[1] + tmp[2] + tmp[3];
    for (; i < n; i++) sum += data[i];
    return sum;
}

// ─── AVX intrinsics (256-bit, 8 × float32) ───

float dot_avx(const float* a, const float* b, int n) {
    __m256 sum_vec = _mm256_setzero_ps();
    int i = 0;
    for (; i + 7 < n; i += 8) {
        __m256 va = _mm256_loadu_ps(a + i);
        __m256 vb = _mm256_loadu_ps(b + i);
        sum_vec = _mm256_add_ps(sum_vec, _mm256_mul_ps(va, vb));
    }
    float tmp[8];
    _mm256_storeu_ps(tmp, sum_vec);
    float sum = 0.0f;
    for (int j = 0; j < 8; j++) sum += tmp[j];
    for (; i < n; i++) sum += a[i] * b[i];
    return sum;
}

float sum_avx(const float* data, int n) {
    __m256 sum_vec = _mm256_setzero_ps();
    int i = 0;
    for (; i + 7 < n; i += 8) {
        __m256 v = _mm256_loadu_ps(data + i);
        sum_vec = _mm256_add_ps(sum_vec, v);
    }
    float tmp[8];
    _mm256_storeu_ps(tmp, sum_vec);
    float sum = 0.0f;
    for (int j = 0; j < 8; j++) sum += tmp[j];
    for (; i < n; i++) sum += data[i];
    return sum;
}

// ─── AVX2 filter: copy src[i] > threshold to dst, return count ───
// Uses AVX2 compare + masked store (compact).

int filter_avx2(const float* src, float* dst, int n, float threshold) {
    __m256 thresh = _mm256_set1_ps(threshold);
    int count = 0;
    int i = 0;
    for (; i + 7 < n; i += 8) {
        __m256 v = _mm256_loadu_ps(src + i);
        __m256 cmp = _mm256_cmp_ps(v, thresh, _CMP_GT_OS);
        unsigned mask = _mm256_movemask_ps(cmp);
        while (mask) {
            unsigned bit = __builtin_ctz(mask);
            dst[count++] = src[i + bit];
            mask &= mask - 1;
        }
    }
    for (; i < n; i++) {
        if (src[i] > threshold) dst[count++] = src[i];
    }
    return count;
}

// ─── AVX-512 dot product (if supported at runtime) ───

#ifdef __AVX512F__
float dot_avx512(const float* a, const float* b, int n) {
    __m512 sum_vec = _mm512_setzero_ps();
    int i = 0;
    for (; i + 15 < n; i += 16) {
        __m512 va = _mm512_loadu_ps(a + i);
        __m512 vb = _mm512_loadu_ps(b + i);
        sum_vec = _mm512_fmadd_ps(va, vb, sum_vec);
    }
    float sum = _mm512_reduce_add_ps(sum_vec);
    for (; i < n; i++) sum += a[i] * b[i];
    return sum;
}
#endif

// ─── Alignment benchmark: aligned vs unaligned load ───

float sum_aligned_load(const float* data, int n) {
    __m256 sum_vec = _mm256_setzero_ps();
    int i = 0;
    for (; i + 7 < n; i += 8) {
        __m256 v = _mm256_load_ps(data + i);
        sum_vec = _mm256_add_ps(sum_vec, v);
    }
    float tmp[8];
    _mm256_storeu_ps(tmp, sum_vec);
    float sum = 0.0f;
    for (int j = 0; j < 8; j++) sum += tmp[j];
    for (; i < n; i++) sum += data[i];
    return sum;
}

// ─── Timing helper ───

template <typename F>
static double bench(const char* label, F func, int iters = ITERS) {
    auto start = std::chrono::high_resolution_clock::now();
    volatile float sink = 0.0f;
    for (int i = 0; i < iters; i++) {
        sink = func();
    }
    (void)sink;
    auto end = std::chrono::high_resolution_clock::now();
    double us = std::chrono::duration<double, std::micro>(end - start).count() / iters;
    std::cout << label << ": " << us << " us/iter" << std::endl;
    return us;
}

int main() {
    std::cout << "=== Vectorization Benchmark ===" << std::endl;
    std::cout << "Array size: " << N << " floats (" << (N * sizeof(float) / 1024) << " KB)" << std::endl;
    std::cout << std::endl;

    float* a = alloc_aligned(N);
    float* b = alloc_aligned(N);
    float* dst = alloc_aligned(N);

    fill_random(a, N);
    fill_random(b, N);

    // ─── Dot Product Benchmarks ───
    std::cout << "--- Dot Product ---" << std::endl;

    float ref = dot_scalar(a, b, N);

    bench("scalar   ", [&]() { return dot_scalar(a, b, N); });
    bench("auto-vec ", [&]() { return dot_auto(a, b, N); });
    bench("SSE      ", [&]() { return dot_sse(a, b, N); });
    bench("AVX      ", [&]() { return dot_avx(a, b, N); });
#ifdef __AVX512F__
    bench("AVX-512  ", [&]() { return dot_avx512(a, b, N); });
#else
    std::cout << "AVX-512  : (not compiled, use -mavx512f)" << std::endl;
#endif

    // ─── Sum Benchmarks ───
    std::cout << std::endl << "--- Sum ---" << std::endl;

    bench("scalar   ", [&]() { return sum_scalar(a, N); });
    bench("auto-vec ", [&]() { return sum_auto(a, N); });
    bench("SSE      ", [&]() { return sum_sse(a, N); });
    bench("AVX      ", [&]() { return sum_avx(a, N); });
    bench("aligned  ", [&]() { return sum_aligned_load(a, N); });

    // ─── Filter Benchmark ───
    std::cout << std::endl << "--- Filter (> 0.0) ---" << std::endl;
    float threshold = 0.0f;

    int cnt_s = 0;
    auto t1 = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERS; i++) {
        cnt_s = filter_scalar(a, dst, N, threshold);
    }
    auto t2 = std::chrono::high_resolution_clock::now();
    double filter_scalar_us = std::chrono::duration<double, std::micro>(t2 - t1).count() / ITERS;
    std::cout << "scalar   : " << filter_scalar_us << " us/iter (" << cnt_s << " kept)" << std::endl;

    int cnt_v = 0;
    auto t3 = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERS; i++) {
        cnt_v = filter_avx2(a, dst, N, threshold);
    }
    auto t4 = std::chrono::high_resolution_clock::now();
    double filter_avx2_us = std::chrono::duration<double, std::micro>(t4 - t3).count() / ITERS;
    std::cout << "AVX2     : " << filter_avx2_us << " us/iter (" << cnt_v << " kept)" << std::endl;

    // ─── Verification ───
    std::cout << std::endl << "--- Verification ---" << std::endl;
    float d_auto = dot_auto(a, b, N);
    float d_sse  = dot_sse(a, b, N);
    float d_avx  = dot_avx(a, b, N);

    bool dot_ok = true;
    if (std::fabs(d_auto - ref) > 1e-3f) { std::cout << "FAIL: auto != scalar" << std::endl; dot_ok = false; }
    if (std::fabs(d_sse - ref) > 1e-3f)  { std::cout << "FAIL: SSE != scalar" << std::endl; dot_ok = false; }
    if (std::fabs(d_avx - ref) > 1e-3f)  { std::cout << "FAIL: AVX != scalar" << std::endl; dot_ok = false; }
    if (dot_ok) std::cout << "All dot product results match (within 1e-3)." << std::endl;

    float s_auto = sum_auto(a, N);
    float s_avx  = sum_avx(a, N);
    if (std::fabs(s_auto - sum_scalar(a, N)) < 1e-2f && std::fabs(s_avx - sum_scalar(a, N)) < 1e-2f) {
        std::cout << "All sum results match (within 1e-2)." << std::endl;
    } else {
        std::cout << "FAIL: sum results differ." << std::endl;
    }

    if (cnt_s == cnt_v) {
        std::cout << "Filter counts match: " << cnt_s << " elements." << std::endl;
    } else {
        std::cout << "FAIL: filter counts differ: scalar=" << cnt_s << " avx2=" << cnt_v << std::endl;
    }

    ALIGNED_FREE(a);
    ALIGNED_FREE(b);
    ALIGNED_FREE(dst);

    return 0;
}

File: phases/15-systems-performance/08-vectorization-in-practice-auto-and-intrinsics/code/main.rs
use std::arch::x86_64::*;
use std::time::Instant;

const N: usize = 1024 * 1024;
const ITERS: usize = 200;

fn alloc_aligned(n: usize, alignment: usize) -> Vec<f32> {
    let layout = std::alloc::Layout::from_size_align(n * std::mem::size_of::<f32>(), alignment).unwrap();
    unsafe {
        let ptr = std::alloc::alloc(layout) as *mut f32;
        for i in 0..n {
            ptr.add(i).write(0.0f32);
        }
        Vec::from_raw_parts(ptr, n, n)
    }
}

fn fill_random(data: &mut [f32]) {
    let mut rng_state: u32 = 12345;
    for i in 0..data.len() {
        rng_state = rng_state.wrapping_mul(1103515245).wrapping_add(12345);
        data[i] = (rng_state as f32) / (u32::MAX as f32) - 0.5;
    }
}

// ─── Scalar implementations ───

fn dot_scalar(a: &[f32], b: &[f32]) -> f32 {
    let mut sum = 0.0f32;
    for i in 0..a.len() {
        sum += a[i] * b[i];
    }
    sum
}

fn sum_scalar(data: &[f32]) -> f32 {
    let mut s = 0.0f32;
    for i in 0..data.len() {
        s += data[i];
    }
    s
}

fn filter_scalar(src: &[f32], dst: &mut [f32], threshold: f32) -> usize {
    let mut count = 0;
    for i in 0..src.len() {
        if src[i] > threshold {
            dst[count] = src[i];
            count += 1;
        }
    }
    count
}

// ─── SSE intrinsics (128-bit, 4 × float32) ───

#[target_feature(enable = "sse")]
unsafe fn dot_sse(a: &[f32], b: &[f32]) -> f32 {
    let mut sum_vec = _mm_setzero_ps();
    let n = a.len();
    let mut i = 0;
    while i + 4 <= n {
        let va = _mm_loadu_ps(a.as_ptr().add(i));
        let vb = _mm_loadu_ps(b.as_ptr().add(i));
        sum_vec = _mm_add_ps(sum_vec, _mm_mul_ps(va, vb));
        i += 4;
    }
    let mut tmp = [0.0f32; 4];
    _mm_storeu_ps(tmp.as_mut_ptr(), sum_vec);
    let mut sum = tmp[0] + tmp[1] + tmp[2] + tmp[3];
    while i < n {
        sum += a[i] * b[i];
        i += 1;
    }
    sum
}

#[target_feature(enable = "sse")]
unsafe fn sum_sse(data: &[f32]) -> f32 {
    let mut sum_vec = _mm_setzero_ps();
    let n = data.len();
    let mut i = 0;
    while i + 4 <= n {
        let v = _mm_loadu_ps(data.as_ptr().add(i));
        sum_vec = _mm_add_ps(sum_vec, v);
        i += 4;
    }
    let mut tmp = [0.0f32; 4];
    _mm_storeu_ps(tmp.as_mut_ptr(), sum_vec);
    let mut sum = tmp[0] + tmp[1] + tmp[2] + tmp[3];
    while i < n {
        sum += data[i];
        i += 1;
    }
    sum
}

// ─── AVX intrinsics (256-bit, 8 × float32) ───

#[target_feature(enable = "avx")]
unsafe fn dot_avx(a: &[f32], b: &[f32]) -> f32 {
    let mut sum_vec = _mm256_setzero_ps();
    let n = a.len();
    let mut i = 0;
    while i + 8 <= n {
        let va = _mm256_loadu_ps(a.as_ptr().add(i));
        let vb = _mm256_loadu_ps(b.as_ptr().add(i));
        sum_vec = _mm256_add_ps(sum_vec, _mm256_mul_ps(va, vb));
        i += 8;
    }
    let mut tmp = [0.0f32; 8];
    _mm256_storeu_ps(tmp.as_mut_ptr(), sum_vec);
    let mut sum = 0.0f32;
    for j in 0..8 {
        sum += tmp[j];
    }
    while i < n {
        sum += a[i] * b[i];
        i += 1;
    }
    sum
}

#[target_feature(enable = "avx")]
unsafe fn sum_avx(data: &[f32]) -> f32 {
    let mut sum_vec = _mm256_setzero_ps();
    let n = data.len();
    let mut i = 0;
    while i + 8 <= n {
        let v = _mm256_loadu_ps(data.as_ptr().add(i));
        sum_vec = _mm256_add_ps(sum_vec, v);
        i += 8;
    }
    let mut tmp = [0.0f32; 8];
    _mm256_storeu_ps(tmp.as_mut_ptr(), sum_vec);
    let mut sum = 0.0f32;
    for j in 0..8 {
        sum += tmp[j];
    }
    while i < n {
        sum += data[i];
        i += 1;
    }
    sum
}

// ─── AVX + FMA intrinsics (fused multiply-add) ───

#[target_feature(enable = "avx2,fma")]
unsafe fn dot_avx_fma(a: &[f32], b: &[f32]) -> f32 {
    let mut sum_vec = _mm256_setzero_ps();
    let n = a.len();
    let mut i = 0;
    while i + 8 <= n {
        let va = _mm256_loadu_ps(a.as_ptr().add(i));
        let vb = _mm256_loadu_ps(b.as_ptr().add(i));
        sum_vec = _mm256_fmadd_ps(va, vb, sum_vec);
        i += 8;
    }
    let mut tmp = [0.0f32; 8];
    _mm256_storeu_ps(tmp.as_mut_ptr(), sum_vec);
    let mut sum = 0.0f32;
    for j in 0..8 {
        sum += tmp[j];
    }
    while i < n {
        sum += a[i] * b[i];
        i += 1;
    }
    sum
}

// ─── AVX2 filter: copy src[i] > threshold to dst, return count ───

#[target_feature(enable = "avx2")]
unsafe fn filter_avx2(src: &[f32], dst: &mut [f32], threshold: f32) -> usize {
    let thresh = _mm256_set1_ps(threshold);
    let n = src.len();
    let mut count: usize = 0;
    let mut i = 0;
    while i + 8 <= n {
        let v = _mm256_loadu_ps(src.as_ptr().add(i));
        let cmp = _mm256_cmp_ps(v, thresh, _CMP_GT_OS);
        let mask = _mm256_movemask_ps(cmp) as u32;
        let mut m = mask;
        while m != 0 {
            let bit = m.trailing_zeros() as usize;
            dst[count] = src[i + bit];
            count += 1;
            m &= m - 1;
        }
        i += 8;
    }
    while i < n {
        if src[i] > threshold {
            dst[count] = src[i];
            count += 1;
        }
        i += 1;
    }
    count
}

// ─── Aligned AVX load benchmark ───

#[target_feature(enable = "avx")]
unsafe fn sum_aligned_avx(data: &[f32]) -> f32 {
    let mut sum_vec = _mm256_setzero_ps();
    let n = data.len();
    let mut i = 0;
    while i + 8 <= n {
        let v = _mm256_load_ps(data.as_ptr().add(i));
        sum_vec = _mm256_add_ps(sum_vec, v);
        i += 8;
    }
    let mut tmp = [0.0f32; 8];
    _mm256_storeu_ps(tmp.as_mut_ptr(), sum_vec);
    let mut sum = 0.0f32;
    for j in 0..8 {
        sum += tmp[j];
    }
    while i < n {
        sum += data[i];
        i += 1;
    }
    sum
}

// ─── Timing helper ───

fn bench<F: Fn() -> f32>(label: &str, func: F) -> f64 {
    let start = Instant::now();
    let mut sink = 0.0f32;
    for _ in 0..ITERS {
        sink = func();
    }
    std::hint::black_box(sink);
    let elapsed = start.elapsed().as_micros() as f64 / ITERS as f64;
    println!("{:<20}: {:.1} us/iter", label, elapsed);
    elapsed
}

fn bench_us<F: Fn() -> usize>(label: &str, func: F) -> f64 {
    let start = Instant::now();
    let mut sink = 0usize;
    for _ in 0..ITERS {
        sink = func();
    }
    std::hint::black_box(sink);
    let elapsed = start.elapsed().as_micros() as f64 / ITERS as f64;
    println!("{:<20}: {:.1} us/iter", label, elapsed);
    elapsed
}

fn main() {
    println!("=== Vectorization Benchmark (Rust) ===");
    println!("Array size: {} floats ({} KB)\n", N, N * std::mem::size_of::<f32>() / 1024);

    let mut a = vec![0.0f32; N];
    let mut b = vec![0.0f32; N];
    let mut dst = vec![0.0f32; N];
    fill_random(&mut a);
    fill_random(&mut b);

    // ─── Dot Product Benchmarks ───
    println!("--- Dot Product ---");
    let ref_val = dot_scalar(&a, &b);

    bench("scalar", || dot_scalar(&a, &b));

    unsafe {
        if is_x86_feature_detected!("sse") {
            bench("SSE", || dot_sse(&a, &b));
        } else {
            println!("SSE                  : (not available)");
        }

        if is_x86_feature_detected!("avx") {
            bench("AVX", || dot_avx(&a, &b));
        } else {
            println!("AVX                  : (not available)");
        }

        if is_x86_feature_detected!("avx2") && is_x86_feature_detected!("fma") {
            bench("AVX+FMA", || dot_avx_fma(&a, &b));
        } else {
            println!("AVX+FMA              : (not available)");
        }
    }

    // ─── Sum Benchmarks ───
    println!("\n--- Sum ---");

    bench("scalar", || sum_scalar(&a));

    unsafe {
        if is_x86_feature_detected!("sse") {
            bench("SSE", || sum_sse(&a));
        }
        if is_x86_feature_detected!("avx") {
            bench("AVX (unaligned)", || sum_avx(&a));
            bench("AVX (aligned)", || sum_aligned_avx(&a));
        }
    }

    // ─── Filter Benchmark ───
    println!("\n--- Filter (> 0.0) ---");
    let threshold = 0.0f32;

    bench_us("scalar", || filter_scalar(&a, &mut dst, threshold));

    unsafe {
        if is_x86_feature_detected!("avx2") {
            bench_us("AVX2", || filter_avx2(&a, &mut dst, threshold));
        } else {
            println!("AVX2                 : (not available)");
        }
    }

    // ─── Verification ───
    println!("\n--- Verification ---");
    let mut all_ok = true;

    let d_auto = dot_scalar(&a, &b);
    if (d_auto - ref_val).abs() > 1e-3 { println!("FAIL: auto != scalar"); all_ok = false; }

    unsafe {
        if is_x86_feature_detected!("sse") {
            let d_sse = dot_sse(&a, &b);
            if (d_sse - ref_val).abs() > 1e-3 { println!("FAIL: SSE != scalar"); all_ok = false; }
        }
        if is_x86_feature_detected!("avx") {
            let d_avx = dot_avx(&a, &b);
            if (d_avx - ref_val).abs() > 1e-3 { println!("FAIL: AVX != scalar"); all_ok = false; }
        }
        if is_x86_feature_detected!("avx2") && is_x86_feature_detected!("fma") {
            let d_fma = dot_avx_fma(&a, &b);
            if (d_fma - ref_val).abs() > 1e-3 { println!("FAIL: FMA != scalar"); all_ok = false; }
        }
    }

    if all_ok { println!("All dot product results match (within 1e-3)."); }

    let s_scalar = sum_scalar(&a);
    unsafe {
        if is_x86_feature_detected!("avx") {
            let s_avx = sum_avx(&a);
            if (s_avx - s_scalar).abs() > 1e-2 { println!("FAIL: sum AVX != scalar"); }
            else { println!("All sum results match (within 1e-2)."); }
        }
    }

    let cnt_scalar = filter_scalar(&a, &mut dst, threshold);
    unsafe {
        if is_x86_feature_detected!("avx2") {
            let cnt_avx2 = filter_avx2(&a, &mut dst, threshold);
            if cnt_scalar == cnt_avx2 {
                println!("Filter counts match: {} elements.", cnt_scalar);
            } else {
                println!("FAIL: filter counts differ: scalar={} avx2={}", cnt_scalar, cnt_avx2);
            }
        }
    }
}